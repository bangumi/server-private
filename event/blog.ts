import { getSlimCacheKey } from '@app/lib/blog/cache';
import redis from '@app/lib/redis.ts';

import { EventOp } from './type';

interface BlogKey {
  entry_id: number;
}

interface Payload {
  op: EventOp;
}

export async function handle(key: string, value: string) {
  const idx = JSON.parse(key) as BlogKey;
  const payload = JSON.parse(value) as Payload;
  switch (payload.op) {
    case EventOp.Create: {
      break;
    }
    case EventOp.Update:
    case EventOp.Delete: {
      await redis.del(getSlimCacheKey(idx.entry_id));
      break;
    }
    case EventOp.Snapshot: {
      break;
    }
  }
}
