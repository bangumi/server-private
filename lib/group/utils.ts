import { db, op, schema } from '@app/drizzle';

/** Is user member of group */
export async function isMemberInGroup(userID: number, groupID: number): Promise<boolean> {
  const [d] = await db
    .select({ uid: schema.chiiGroupMembers.uid, gid: schema.chiiGroupMembers.gid })
    .from(schema.chiiGroupMembers)
    .where(
      op.and(
        op.eq(schema.chiiGroupMembers.uid, userID),
        op.eq(schema.chiiGroupMembers.gid, groupID),
      ),
    )
    .limit(1);

  return Boolean(d);
}
