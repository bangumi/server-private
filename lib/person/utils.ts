import { db, op, schema } from '@app/drizzle';

import type { PersonCat } from './type';

export async function getPersonCollect(
  cat: PersonCat,
  userID: number,
  personID: number,
): Promise<number | null> {
  const [d] = await db
    .select()
    .from(schema.chiiPersonCollects)
    .where(
      op.and(
        op.eq(schema.chiiPersonCollects.cat, cat),
        op.eq(schema.chiiPersonCollects.uid, userID),
        op.eq(schema.chiiPersonCollects.mid, personID),
      ),
    )
    .limit(1);
  if (!d) {
    return null;
  }
  return d.createdAt;
}
