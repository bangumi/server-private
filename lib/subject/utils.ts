import * as php from '@trim21/php-serialize';
import { DateTime } from 'luxon';

import { db, decr, incr, op, type orm, schema, type Txn } from '@app/drizzle';

import { type CollectionType, SubjectType, type UserEpisodeStatusItem } from './type';
import { EpisodeCollectionStatus, getCollectionTypeField } from './type';

/** 更新条目收藏计数，需要在事务中执行 */
export async function updateSubjectCollection(
  t: Txn,
  subjectID: number,
  newType: CollectionType,
  oldType?: CollectionType,
) {
  if (oldType && oldType === newType) {
    return;
  }
  const toUpdate: Record<string, op.SQL> = {};
  toUpdate[getCollectionTypeField(newType)] = incr(
    schema.chiiSubjects[getCollectionTypeField(newType)],
  );
  if (oldType) {
    toUpdate[getCollectionTypeField(oldType)] = decr(
      schema.chiiSubjects[getCollectionTypeField(oldType)],
    );
  }
  await t
    .update(schema.chiiSubjects)
    .set(toUpdate)
    .where(op.eq(schema.chiiSubjects.id, subjectID))
    .limit(1);
}

function getRatingField(rate: number) {
  return [
    undefined,
    'rate1',
    'rate2',
    'rate3',
    'rate4',
    'rate5',
    'rate6',
    'rate7',
    'rate8',
    'rate9',
    'rate10',
  ][rate];
}

/** 更新条目评分，需要在事务中执行 */
export async function updateSubjectRating(
  t: Txn,
  subjectID: number,
  oldRate: number,
  newRate: number,
) {
  if (oldRate === newRate) {
    return;
  }
  const newField = getRatingField(newRate);
  const oldField = getRatingField(oldRate);
  const toUpdate: Record<string, op.SQL> = {};
  if (newField) {
    const field = newField as keyof orm.ISubjectFields;
    toUpdate[field] = incr(schema.chiiSubjectFields[field]);
  }
  if (oldField) {
    const field = oldField as keyof orm.ISubjectFields;
    toUpdate[field] = decr(schema.chiiSubjectFields[field]);
  }
  if (Object.keys(toUpdate).length === 0) {
    return;
  }
  await t
    .update(schema.chiiSubjectFields)
    .set(toUpdate)
    .where(op.eq(schema.chiiSubjectFields.id, subjectID))
    .limit(1);
}

/** 更新条目剧集进度，需要在事务中执行 */
export async function updateSubjectEpisodeProgress(
  t: Txn,
  userID: number,
  subjectID: number,
  episodeID: number,
  type: number,
): Promise<number> {
  const [current] = await t
    .select()
    .from(schema.chiiEpStatus)
    .where(
      op.and(op.eq(schema.chiiEpStatus.uid, userID), op.eq(schema.chiiEpStatus.sid, subjectID)),
    );
  let watchedEpisodes = 0;
  if (current?.status) {
    const epStatusList = parseSubjectEpStatus(current.status);
    epStatusList[episodeID] = {
      eid: episodeID,
      type,
    };
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t
      .update(schema.chiiEpStatus)
      .set({ status: newStatus, updatedAt: DateTime.now().toUnixInteger() })
      .where(op.eq(schema.chiiEpStatus.id, current.id))
      .limit(1);
  } else {
    const epStatusList: Record<number, UserEpisodeStatusItem> = {};
    epStatusList[episodeID] = {
      eid: episodeID,
      type,
    };
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t.insert(schema.chiiEpStatus).values({
      uid: userID,
      sid: subjectID,
      status: newStatus,
      updatedAt: DateTime.now().toUnixInteger(),
    });
  }
  return watchedEpisodes;
}

/** 标记条目剧集为已观看，需要在事务中执行 */
export async function markEpisodesAsWatched(
  t: Txn,
  userID: number,
  subjectID: number,
  episodeIDs: number[],
  revertOthers = false,
): Promise<number> {
  const epStatusList: Record<number, UserEpisodeStatusItem> = {};
  for (const episodeID of episodeIDs) {
    epStatusList[episodeID] = {
      eid: episodeID,
      type: EpisodeCollectionStatus.Done,
    };
  }
  const [current] = await t
    .select()
    .from(schema.chiiEpStatus)
    .where(
      op.and(op.eq(schema.chiiEpStatus.uid, userID), op.eq(schema.chiiEpStatus.sid, subjectID)),
    );
  let watchedEpisodes = 0;
  if (current?.status) {
    const oldList = parseSubjectEpStatus(current.status);
    for (const x of Object.values(oldList)) {
      if (episodeIDs.includes(x.eid)) {
        continue;
      }
      if (revertOthers && x.type === EpisodeCollectionStatus.Done) {
        epStatusList[x.eid] = {
          eid: x.eid,
          type: EpisodeCollectionStatus.None,
        };
      } else {
        epStatusList[x.eid] = x;
      }
    }
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t
      .update(schema.chiiEpStatus)
      .set({ status: newStatus, updatedAt: DateTime.now().toUnixInteger() })
      .where(op.eq(schema.chiiEpStatus.id, current.id))
      .limit(1);
  } else {
    watchedEpisodes = Object.values(epStatusList).filter(
      (x) => x.type === EpisodeCollectionStatus.Done,
    ).length;
    const newStatus = php.stringify(epStatusList);
    await t.insert(schema.chiiEpStatus).values({
      uid: userID,
      sid: subjectID,
      status: newStatus,
      updatedAt: DateTime.now().toUnixInteger(),
    });
  }
  return watchedEpisodes;
}

export function parseSubjectEpStatus(status: string): Record<number, UserEpisodeStatusItem> {
  const result: Record<number, UserEpisodeStatusItem> = {};
  if (!status) {
    return result;
  }
  const epStatusList = php.parse(status) as Record<number, UserEpisodeStatusItem>;
  for (const x of Object.values(epStatusList)) {
    result[x.eid] = x;
  }
  return result;
}

export async function getEpStatus(
  userID: number,
  subjectID: number,
): Promise<Record<number, UserEpisodeStatusItem>> {
  const [data] = await db
    .select()
    .from(schema.chiiEpStatus)
    .where(
      op.and(op.eq(schema.chiiEpStatus.uid, userID), op.eq(schema.chiiEpStatus.sid, subjectID)),
    )
    .limit(1);
  if (!data) {
    return {};
  }
  return parseSubjectEpStatus(data.status);
}

/** 完成条目进度，需要在事务中执行 */
export async function completeSubjectProgress(
  t: Txn,
  userID: number,
  subject: orm.ISubject,
  interest: Partial<orm.ISubjectInterest>,
) {
  if (subject.eps > 0) {
    interest.epStatus = subject.eps;
  }
  if (subject.volumes > 0) {
    interest.volStatus = subject.volumes;
  }
  if ([SubjectType.Anime, SubjectType.Real].includes(subject.typeID)) {
    const episodes = await t
      .select({ id: schema.chiiEpisodes.id })
      .from(schema.chiiEpisodes)
      .where(
        op.and(
          op.eq(schema.chiiEpisodes.subjectID, subject.id),
          op.eq(schema.chiiEpisodes.type, 0),
          op.eq(schema.chiiEpisodes.ban, 0),
        ),
      );
    const episodeIDs = episodes.map((e) => e.id);
    if (episodeIDs.length > 0) {
      await markEpisodesAsWatched(t, userID, subject.id, episodeIDs);
    }
  }
}
