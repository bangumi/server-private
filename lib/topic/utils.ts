import type * as orm from '@app/drizzle/orm.ts';
import { HotGroups } from '@app/lib/group/consts.ts';

export function scoredUpdateTime(timestamp: number, topic: orm.IGroupTopic): number {
  if (HotGroups.includes(topic.gid) && topic.replies > 0) {
    const createdAt = topic.createdAt;
    const createdHours = (timestamp - createdAt) / 3600;
    const gravity = 1.8;
    const baseScore = (Math.pow(createdHours + 0.1, gravity) / topic.replies) * 200;
    const scoredLastPost = Math.trunc(timestamp - baseScore);
    return Math.min(scoredLastPost, timestamp);
  }

  return timestamp;
}
