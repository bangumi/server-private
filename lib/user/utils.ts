import { db, op, schema } from '@app/drizzle';
import { UnexpectedNotFoundError } from '@app/lib/error.ts';
import redis from '@app/lib/redis.ts';
import * as fetcher from '@app/lib/types/fetcher.ts';
import type * as res from '@app/lib/types/res.ts';
import {
  getFollowersCacheKey,
  getFriendsCacheKey,
  getRelationCacheKey,
} from '@app/lib/user/cache.ts';

/** Cached: Get friend ids of user(uid) */
export async function fetchFriends(uid?: number): Promise<number[]> {
  if (!uid) {
    return [];
  }

  const cached = await redis.get(getFriendsCacheKey(uid));
  if (cached) {
    return JSON.parse(cached) as number[];
  }

  const friends = await db
    .select({ fid: schema.chiiFriends.fid })
    .from(schema.chiiFriends)
    .where(op.eq(schema.chiiFriends.uid, uid));
  const result = friends.map((x) => x.fid);
  await redis.setex(getFriendsCacheKey(uid), 3600, JSON.stringify(result));
  return result;
}

/** Cached: Get follower ids of user(uid) */
export async function fetchFollowers(uid?: number): Promise<number[]> {
  if (!uid) {
    return [];
  }

  const cached = await redis.get(getFollowersCacheKey(uid));
  if (cached) {
    return JSON.parse(cached) as number[];
  }

  const followers = await db
    .select({ uid: schema.chiiFriends.uid })
    .from(schema.chiiFriends)
    .where(op.eq(schema.chiiFriends.fid, uid));
  const result = followers.map((x) => x.uid);
  await redis.setex(getFollowersCacheKey(uid), 3600, JSON.stringify(result));
  return result;
}

/** Cached: Is user(another) is friend of user(uid) */
export async function isFriends(uid: number, another: number): Promise<boolean> {
  const cached = await redis.get(getRelationCacheKey(uid, another));
  if (cached) {
    return cached === '1';
  }

  const [d] = await db
    .select({ uid: schema.chiiFriends.uid, fid: schema.chiiFriends.fid })
    .from(schema.chiiFriends)
    .where(op.and(op.eq(schema.chiiFriends.uid, uid), op.eq(schema.chiiFriends.fid, another)))
    .limit(1);
  const result = d ? 1 : 0;
  await redis.setex(getRelationCacheKey(uid, another), 3600, result);
  return result === 1;
}

export function ghostUser(uid: number): res.ISlimUser {
  return {
    id: 0,
    username: uid.toString(),
    nickname: `deleted or missing user ${uid}`,
    avatar: {
      small: '',
      medium: '',
      large: '',
    },
    group: 0,
    sign: '',
    joinedAt: 0,
  };
}

/** Cached */
export async function fetchUserX(uid: number): Promise<res.ISlimUser> {
  const user = await fetcher.fetchSlimUserByID(uid);
  if (!user) {
    throw new UnexpectedNotFoundError(`user ${uid} not found`);
  }
  return user;
}
