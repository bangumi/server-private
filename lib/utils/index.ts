import * as crypto from 'node:crypto';

import { customAlphabet } from 'nanoid/async';

const base62Chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
if (base62Chars.length !== 62) {
  throw new TypeError('characters is not 62 length');
}

const generator = customAlphabet(base62Chars, 32);

export const randomBase62String = (size: number) => generator(size);

export async function randomBytes(size: number): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    crypto.randomBytes(size, (err, buf) => {
      if (err) {
        reject(err);
      } else {
        resolve(buf);
      }
    });
  });
}

/**
 * Parse string as int, strictly
 *
 * 避免出现 `parseInt('1abc') === 1` 的问题
 */
export function intval(value: string | number): number {
  const v = parseIntStrict(value);
  if (v === undefined) {
    throw new Error(`"${value}" is not a valid integer`);
  }
  return v;
}

function parseIntStrict(integer: string | number) {
  if (typeof integer === 'number') {
    return Number.isInteger(integer) ? integer : undefined;
  }

  const n = Number(integer);

  if (Number.isInteger(n)) {
    return n;
  }
}

export function md5(s: string) {
  return crypto.createHash('md5').update(s).digest('hex');
}
