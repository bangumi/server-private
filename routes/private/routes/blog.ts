import { Type as t } from '@sinclair/typebox';

import { db, op } from '@app/drizzle/db.ts';
import * as schema from '@app/drizzle/schema';
import { NotFoundError } from '@app/lib/error.ts';
import { Security, Tag } from '@app/lib/openapi/index.ts';
import * as convert from '@app/lib/types/convert.ts';
import * as fetcher from '@app/lib/types/fetcher.ts';
import * as res from '@app/lib/types/res.ts';
import { isFriends } from '@app/lib/user/utils.ts';
import type { App } from '@app/routes/type.ts';

// eslint-disable-next-line @typescript-eslint/require-await
export async function setup(app: App) {
  app.get(
    '/blogs/:entryID',
    {
      schema: {
        summary: '获取日志详情',
        operationId: 'getBlogEntry',
        tags: [Tag.Blog],
        security: [{ [Security.HTTPBearer]: [] }],
        params: t.Object({
          entryID: t.Integer(),
        }),
        response: {
          200: res.Ref(res.BlogEntry),
        },
      },
    },
    async ({ auth, params: { entryID } }) => {
      const [item] = await db
        .select()
        .from(schema.chiiBlogEntries)
        .innerJoin(schema.chiiUsers, op.eq(schema.chiiUsers.id, schema.chiiBlogEntries.uid))
        .where(op.eq(schema.chiiBlogEntries.id, entryID));
      if (!item) {
        throw new NotFoundError('Blog entry not found');
      }
      const entry = convert.toBlogEntry(item.chii_blog_entry, item.chii_members);
      const isFriend = await isFriends(entry.user.id, auth.userID);
      if (!entry.public && entry.user.id !== auth.userID && !isFriend) {
        throw new NotFoundError('Blog entry not found');
      }
      return entry;
    },
  );

  app.get(
    '/blogs/:entryID/subjects',
    {
      schema: {
        summary: '获取日志的关联条目',
        operationId: 'getBlogRelatedSubjects',
        tags: [Tag.Blog],
        security: [{ [Security.HTTPBearer]: [] }],
        params: t.Object({
          entryID: t.Integer(),
        }),
        response: {
          200: t.Array(res.Ref(res.SlimSubject)),
        },
      },
    },
    async ({ auth, params: { entryID } }) => {
      const entry = await fetcher.fetchSlimBlogEntryByID(entryID);
      if (!entry) {
        throw new NotFoundError('Blog entry not found');
      }
      const isFriend = await isFriends(entry.uid, auth.userID);
      if (!entry.public && entry.uid !== auth.userID && !isFriend) {
        throw new NotFoundError('Blog entry not found');
      }

      const data = await db
        .select({ id: schema.chiiSubjectRelatedBlogs.subjectID })
        .from(schema.chiiSubjectRelatedBlogs)
        .where(op.eq(schema.chiiSubjectRelatedBlogs.entryID, entryID))
        .orderBy(op.desc(schema.chiiSubjectRelatedBlogs.id));
      const subjectIDs = data.map((item) => item.id);
      const subjects = await fetcher.fetchSlimSubjectsByIDs(subjectIDs);
      const result = [];
      for (const id of subjectIDs) {
        const subject = subjects[id];
        if (!subject) {
          continue;
        }
        if (!auth.allowNsfw && subject.nsfw) {
          continue;
        }
        result.push(subject);
      }
      return result;
    },
  );

  app.get(
    '/blogs/:entryID/photos',
    {
      schema: {
        summary: '获取日志的图片',
        operationId: 'getBlogPhotos',
        tags: [Tag.Blog],
        security: [{ [Security.HTTPBearer]: [] }],
        params: t.Object({
          entryID: t.Integer(),
        }),
        querystring: t.Object({
          limit: t.Optional(
            t.Integer({ default: 20, minimum: 1, maximum: 100, description: 'max 100' }),
          ),
          offset: t.Optional(t.Integer({ default: 0, minimum: 0, description: 'min 0' })),
        }),
        response: {
          200: res.Paged(res.Ref(res.BlogPhoto)),
        },
      },
    },
    async ({ auth, params: { entryID }, query: { limit = 20, offset = 0 } }) => {
      const entry = await fetcher.fetchSlimBlogEntryByID(entryID);
      if (!entry) {
        throw new NotFoundError('Blog entry not found');
      }
      // 只允许查看自己的日志图片
      if (entry.uid !== auth.userID) {
        throw new NotFoundError('Blog entry not found');
      }

      const [{ count = 0 } = {}] = await db
        .select({ count: op.count() })
        .from(schema.chiiBlogPhotos)
        .where(op.eq(schema.chiiBlogPhotos.eid, entryID));

      const data = await db
        .select()
        .from(schema.chiiBlogPhotos)
        .where(op.eq(schema.chiiBlogPhotos.eid, entryID))
        .orderBy(op.desc(schema.chiiBlogPhotos.createdAt))
        .limit(limit)
        .offset(offset);

      const photos = data.map((photo) => convert.toBlogPhoto(photo));

      return {
        data: photos,
        total: count,
      };
    },
  );
}
