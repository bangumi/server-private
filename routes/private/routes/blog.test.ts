import { describe, expect, test } from 'vitest';

import { emptyAuth } from '@app/lib/auth/index.ts';
import { createTestServer } from '@app/tests/utils.ts';

import { setup } from './blog.ts';

describe('blog', () => {
  const testUID = 287622;
  const publicEntryID = 319484;
  const privateEntryID = 319486;

  test('should get blog entry', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: `/blogs/${publicEntryID}`,
    });
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should not get private blog entry from other user', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUID + 1, // different user
      },
    });
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: `/blogs/${privateEntryID}`,
    });
    expect(res.statusCode).toBe(404);
  });

  test('should get private blog entry from friend', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 427613, // friend
      },
    });
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: `/blogs/${privateEntryID}`,
    });
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should get private blog entry from self', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUID, // same user
      },
    });
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: `/blogs/${privateEntryID}`,
    });
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should get blog related subjects', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: `/blogs/${publicEntryID}/subjects`,
    });
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should get blog photos', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUID, // same user
      },
    });
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: `/blogs/${publicEntryID}/photos`,
      query: { limit: '2', offset: '0' },
    });
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should not get blog photos from other user', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUID + 1, // other user
      },
    });
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: `/blogs/${privateEntryID}/photos`,
    });
    expect(res.statusCode).toBe(404);
  });
});
