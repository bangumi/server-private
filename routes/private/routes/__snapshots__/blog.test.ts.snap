// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`blog > should get blog entry 1`] = `
Object {
  "content": "试试",
  "createdAt": 1680065731,
  "icon": "https://lain.bgm.tv/pic/photo/g/no_photo.png",
  "id": 319484,
  "noreply": 0,
  "public": true,
  "related": 1,
  "replies": 1,
  "tags": Array [
    "动画",
  ],
  "title": "试试",
  "type": 1,
  "updatedAt": 1680072373,
  "user": Object {
    "avatar": Object {
      "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
      "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
      "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
    },
    "group": 0,
    "id": 287622,
    "joinedAt": 0,
    "nickname": "nickname 287622",
    "sign": "sing 287622",
    "username": "287622",
  },
  "views": 0,
}
`;

exports[`blog > should get blog photos 1`] = `
Object {
  "data": Array [],
  "total": 0,
}
`;

exports[`blog > should get blog related subjects 1`] = `
Array [
  Object {
    "id": 184017,
    "info": "233话 / 1582-10-14",
    "locked": false,
    "name": "沙盒",
    "nameCN": "Bangumi Wiki 动画测试用沙盘",
    "nsfw": false,
    "rating": Object {
      "count": Array [
        1,
        1,
        0,
        0,
        0,
        1,
        1,
        0,
        0,
        3,
      ],
      "rank": 0,
      "score": 6.57,
      "total": 7,
    },
    "type": 2,
  },
]
`;

exports[`blog > should get private blog entry from friend 1`] = `
Object {
  "content": "感谢 @ klion 提供数据，相关条目数据进行了一次更新。


[img]https://s3.bgm38.com/images/Snipaste_2023-06-05_16-33-39.jpg[/img]

",
  "createdAt": 1680065928,
  "icon": "https://lain.bgm.tv/pic/photo/g/no_photo.png",
  "id": 319486,
  "noreply": 0,
  "public": false,
  "related": 1,
  "replies": 1,
  "tags": Array [
    "动画",
  ],
  "title": "[2023-06-05] 相关条目数据更新",
  "type": 1,
  "updatedAt": 1724513422,
  "user": Object {
    "avatar": Object {
      "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
      "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
      "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
    },
    "group": 0,
    "id": 287622,
    "joinedAt": 0,
    "nickname": "nickname 287622",
    "sign": "sing 287622",
    "username": "287622",
  },
  "views": 0,
}
`;

exports[`blog > should get private blog entry from self 1`] = `
Object {
  "content": "感谢 @ klion 提供数据，相关条目数据进行了一次更新。


[img]https://s3.bgm38.com/images/Snipaste_2023-06-05_16-33-39.jpg[/img]

",
  "createdAt": 1680065928,
  "icon": "https://lain.bgm.tv/pic/photo/g/no_photo.png",
  "id": 319486,
  "noreply": 0,
  "public": false,
  "related": 1,
  "replies": 1,
  "tags": Array [
    "动画",
  ],
  "title": "[2023-06-05] 相关条目数据更新",
  "type": 1,
  "updatedAt": 1724513422,
  "user": Object {
    "avatar": Object {
      "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
      "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
      "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
    },
    "group": 0,
    "id": 287622,
    "joinedAt": 0,
    "nickname": "nickname 287622",
    "sign": "sing 287622",
    "username": "287622",
  },
  "views": 0,
}
`;
